src/server/service/kv.rs

```diff
@@ -343,10 +507,240 @@ impl<E: Engine, L: LockMgr> Batcher<E, L> for ReadBatcher {
     }
 }
 
+#[derive(Hash, PartialEq, Eq, Debug)]
+struct WriteId {
+    region_id: RegionVerId,
+    sync_log: bool,
+}
+
+impl WriteId {
+    fn from_context(ctx: &Context) -> Self {
+        WriteId {
+            region_id: RegionVerId::from_context(ctx),
+            sync_log: ctx.get_sync_log(),
+        }
+    }
+}
+
+struct WriteBatch {
+    command: Command,
+    key_set: HashSet<Vec<u8>>,
+    write_bytes: usize,
+}
+
+impl Default for WriteBatch {
+    fn default() -> Self {
+        WriteBatch {
+            command: Command::Batch {
+                ids: Vec::new(),
+                commands: Vec::new(),
+            },
+            key_set: HashSet::default(),
+            write_bytes: 0,
+        }
+    }
+}
+
+struct WriteBatcher {
+    cmd: BatchableRequestKind,
+    router: HashMap<WriteId, WriteBatch>,
+}
+
+impl WriteBatcher {
+    fn new(cmd: BatchableRequestKind) -> Self {
+        WriteBatcher {
+            cmd,
+            router: HashMap::default(),
+        }
+    }
+
+    fn is_batchable_context(ctx: &Context) -> bool {
+        storage::is_normal_priority(ctx.get_priority())
+    }
+
+    fn is_batchable_prewrite(req: &PrewriteRequest) -> bool {
+        // only batch optimistic prewrite
+        req.get_for_update_ts() == 0 && Self::is_batchable_context(req.get_context())
+    }
+}
+
+impl<E: Engine, L: LockMgr> Batcher<E, L> for WriteBatcher {
+    fn filter(
+        &mut self,
+        request_id: u64,
+        request: &mut batch_commands_request::request::Cmd,
+    ) -> bool {
+        // TODO(tabokie): lazy initialization of keys
+        let (ver, keys): (_, Vec<Key>) = match request {
+            batch_commands_request::request::Cmd::Prewrite(req)
+                if self.cmd == BatchableRequestKind::Prewrite
+                    && Self::is_batchable_prewrite(req) =>
```

Comment:
If  size of the request is too large (such as mutations.len() > 16. ) , could we skip it and make the request send more quickly?